crates/polars-core/src/series/arithmetic/list_borrowed.rs

```diff
@@ -1,177 +1,1026 @@
     }
 }
 
-/// Given an ArrayRef, return true if it's a LargeListArrays and it has one or
-/// more nulls.
-fn does_list_have_nulls(data: &ArrayRef) -> bool {
-    if let Some(list_chunk) = data.as_any().downcast_ref::<LargeListArray>() {
-        if list_chunk
-            .validity()
-            .map(|bitmap| bitmap.unset_bits() > 0)
-            .unwrap_or(false)
-        {
-            true
-        } else {
-            does_list_have_nulls(list_chunk.values())
+#[derive(Debug, Clone)]
+pub enum NumericListOp {
+    Add,
+    Sub,
+    Mul,
+    Div,
+    Rem,
+    FloorDiv,
+}
+
+impl NumericListOp {
+    fn name(&self) -> &'static str {
+        match self {
+            Self::Add => "add",
+            Self::Sub => "sub",
+            Self::Mul => "mul",
+            Self::Div => "div",
+            Self::Rem => "rem",
+            Self::FloorDiv => "floor_div",
         }
-    } else {
-        false
     }
-}
 
-/// Return whether the left and right have the same shape. We assume neither has
-/// any nulls, recursively.
-fn lists_same_shapes(left: &ArrayRef, right: &ArrayRef) -> bool {
-    debug_assert!(!does_list_have_nulls(left));
-    debug_assert!(!does_list_have_nulls(right));
-    let left_as_list = left.as_any().downcast_ref::<LargeListArray>();
-    let right_as_list = right.as_any().downcast_ref::<LargeListArray>();
-    match (left_as_list, right_as_list) {
-        (Some(left), Some(right)) => {
-            left.offsets() == right.offsets() && lists_same_shapes(left.values(), right.values())
-        },
-        (None, None) => left.len() == right.len(),
-        _ => false,
+    pub fn try_get_leaf_supertype(
+        &self,
+        prim_dtype_lhs: &DataType,
+        prim_dtype_rhs: &DataType,
+    ) -> PolarsResult<DataType> {
+        let dtype = try_get_supertype(prim_dtype_lhs, prim_dtype_rhs)?;
+
+        Ok(if matches!(self, Self::Div) {
+            if dtype.is_float() {
+                dtype
+            } else {
+                DataType::Float64
+            }
+        } else {
+            dtype
+        })
+    }
+
+    pub fn execute(&self, lhs: &Series, rhs: &Series) -> PolarsResult<Series> {
+        // Ideally we only need to rechunk the leaf array, but getting the
+        // list offsets of a ListChunked triggers a rechunk anyway, so we just
+        // do it here.
+        let lhs = lhs.rechunk();
+        let rhs = rhs.rechunk();
+
+        let binary_op_exec = match BinaryListNumericOpHelper::try_new(
+            self.clone(),
+            lhs.name().clone(),
+            lhs.dtype(),
+            rhs.dtype(),
+            lhs.len(),
+            rhs.len(),
+            {
+                let (a, b) = lhs.list_offsets_and_validities_recursive();
+                (a, b, lhs.clone())
+            },
+            {
+                let (a, b) = rhs.list_offsets_and_validities_recursive();
+                (a, b, rhs.clone())
+            },
+            lhs.rechunk_validity(),
```

Comment:
Why do we have to rechunk_validity again after rechunking before?

Reply:
It won't rechunk because we already did, but that function is the easiest way to get the validity from a `dyn SeriesTrait`. It's just named `rechunk_validity()` to indicate to the caller that it could potentially perform a rechunk.

If you want I can change it to `lhs.chunks()[0].validity()`, but I think I prefer it as it is right now